File: in_memory_cache_service/src/tests/components/note_component.test.ts
import {
  describe,
  beforeAll,
  beforeEach,
  afterAll,
  test,
  expect,
} from "@jest/globals";
import InMemoryDatabaseFactory from "../../domains/in_memory_database/in_memory_database_factory";
import {
  IN_MEMORY_DATABASE_HOST_NAME,
  IN_MEMORY_DATABASE_HOST_PORT,
} from "../../constants/application_constants";
import NoteActiveRecord from "../../domains/note/note_active_record";
import UserRepository from "../../domains/user/user_repository";
import UserRepositoryFactory from "../../domains/user/user_repository_factory";
import UserEntityFactory from "../../domains/user/user_entity_factory";
import {
  USER_ID,
  USER_USERNAME,
} from "../../constants/domains/user_constants";
import {
  NOT_EXISTING_NOTE_ERROR_MESSAGE,
  NOT_EXISTING_NOTE_ERROR_NAME,
  NOTE_BODY,
  NOTE_CREATED_AT,
  NOTE_ID,
  NOTE_TITLE,
  NOTE_UPDATED_AT,
} from "../../constants/domains/note_constants";
import NoteEntityFactory from "../../domains/note/note_entity_factory";
import NoteRepository from "../../domains/note/note_repository";
import NoteActiveRecordDecorator from "../../domains/note/note_active_record_decorator";

describe("Test Component Note Behavior", () => {
  let userRepository: UserRepository;
  let noteActiveRecord: NoteActiveRecord;
  let noteRepository: NoteRepository;

  beforeAll(async () => {
    const inMemoryDatabase = await InMemoryDatabaseFactory.getInstance(
      IN_MEMORY_DATABASE_HOST_NAME,
      IN_MEMORY_DATABASE_HOST_PORT,
    );

    userRepository = await UserRepositoryFactory.getInstance();

    noteActiveRecord = new NoteActiveRecordDecorator(inMemoryDatabase);

    noteRepository = new NoteRepository(noteActiveRecord);

    try {
      const user = UserEntityFactory.getInstance(USER_ID, USER_USERNAME);

      await userRepository.createUser(user);
      // eslint-disable-next-line @typescript-eslint/no-unused-vars, no-empty
    } catch (error) {}
  });

  beforeEach(async () => {
    try {
      await noteActiveRecord.deleteNoteOnInMemoryDatabase(NOTE_ID, USER_ID);
      // eslint-disable-next-line @typescript-eslint/no-unused-vars, no-empty
    } catch (error) {}
  });

  afterAll(async () => {
    try {
      await userRepository.deleteUser(USER_ID);
      // eslint-disable-next-line @typescript-eslint/no-unused-vars, no-empty
    } catch (error) {}
  });

  test('Test Fetching All Users\'s Notes', async () => {
    const note = NoteEntityFactory.getInstance(
      NOTE_ID,
      NOTE_TITLE,
      NOTE_BODY,
      NOTE_CREATED_AT,
      NOTE_UPDATED_AT,
      USER_ID,
    );

    await noteActiveRecord.createNoteOnInMemoryDatabase(note);

    const userNotes = await noteRepository.getNotes(USER_ID);

    expect(userNotes[0].id).toEqual(NOTE_ID);
    expect(userNotes[0].title).toEqual(NOTE_TITLE);
    expect(userNotes[0].body).toEqual(NOTE_BODY);
    expect(userNotes[0].createdAt).toEqual(NOTE_CREATED_AT);
    expect(userNotes[0].updatedAt).toEqual(NOTE_UPDATED_AT);
    expect(userNotes[0].userId).toEqual(USER_ID);
  });

  test('Test Fetching Users\'s Note', async () => {
    const note = NoteEntityFactory.getInstance(
      NOTE_ID,
      NOTE_TITLE,
      NOTE_BODY,
      NOTE_CREATED_AT,
      NOTE_UPDATED_AT,
      USER_ID,
    );

    await noteActiveRecord.createNoteOnInMemoryDatabase(note);

    const userNote = await noteRepository.getNote(NOTE_ID);

    expect(userNote.id).toEqual(NOTE_ID);
    expect(userNote.title).toEqual(NOTE_TITLE);
    expect(userNote.body).toEqual(NOTE_BODY);
    expect(userNote.createdAt).toEqual(NOTE_CREATED_AT);
    expect(userNote.updatedAt).toEqual(NOTE_UPDATED_AT);
    expect(userNote.userId).toEqual(USER_ID);
  });

  test('Test Creating Note On Database', async () => {
    const note = NoteEntityFactory.getInstance(
      NOTE_ID,
      NOTE_TITLE,
      NOTE_BODY,
      NOTE_CREATED_AT,
      NOTE_UPDATED_AT,
      USER_ID,
    );

    await noteRepository.createNote(note);

    const userNote =
      await noteActiveRecord.getNoteFromInMemoryDatabase(NOTE_ID);

    expect(userNote.id).toEqual(NOTE_ID);
    expect(userNote.title).toEqual(NOTE_TITLE);
    expect(userNote.body).toEqual(NOTE_BODY);
    expect(userNote.createdAt).toEqual(NOTE_CREATED_AT);
    expect(userNote.updatedAt).toEqual(NOTE_UPDATED_AT);
    expect(userNote.userId).toEqual(USER_ID);
  });

  test('Test Updating Note On Database', async () => {
    const note = NoteEntityFactory.getInstance(
      NOTE_ID,
      "",
      "",
      Math.random(),
      Math.random(),
      USER_ID,
    );

    await noteActiveRecord.createNoteOnInMemoryDatabase(note);

    const noteWithUpdatedData = NoteEntityFactory.getInstance(
      NOTE_ID,
      NOTE_TITLE,
      NOTE_BODY,
      NOTE_CREATED_AT,
      NOTE_UPDATED_AT,
      USER_ID,
    );

    await noteRepository.updateNote(noteWithUpdatedData);

    const updatedNote =
      await noteActiveRecord.getNoteFromInMemoryDatabase(NOTE_ID);

    expect(updatedNote.id).toEqual(NOTE_ID);
    expect(updatedNote.title).toEqual(NOTE_TITLE);
    expect(updatedNote.body).toEqual(NOTE_BODY);
    expect(updatedNote.createdAt).toEqual(NOTE_CREATED_AT);
    expect(updatedNote.updatedAt).toEqual(NOTE_UPDATED_AT);
    expect(updatedNote.userId).toEqual(USER_ID);
  });

  test('Test Deleting Note On Database', async () => {
    const note = NoteEntityFactory.getInstance(
      NOTE_ID,
      NOTE_TITLE,
      NOTE_BODY,
      NOTE_CREATED_AT,
      NOTE_UPDATED_AT,
      USER_ID,
    );

    await noteActiveRecord.createNoteOnInMemoryDatabase(note);

    try {
      await noteRepository.deleteNote(NOTE_ID, USER_ID);
    } catch (error) {
      const castedError = error as Error;

      expect(castedError.name).toEqual(NOT_EXISTING_NOTE_ERROR_NAME);
      expect(castedError.message).toEqual(NOT_EXISTING_NOTE_ERROR_MESSAGE);
    }
  });
});

File: data_source_service/test/components/note_component_test.rb
require "test_helper"

class NoteComponentTest < ActiveSupport::TestCase
  test "Test Fetching Note From Database" do
    User.create id: UserConstants::USER_TO_BE_CREATED_ID, username: UserConstants::USER_TO_BE_CREATED_USERNAME

    Note.create id: NoteConstants::NOTE_TO_BE_CREATED_ID,
      title: NoteConstants::NOTE_TO_BE_CREATED_TITLE,
      body: NoteConstants::NOTE_TO_BE_CREATED_BODY,
      created_at: NoteConstants::NOTE_TO_BE_CREATED_CREATED_AT_TIMESTAMP,
      updated_at: NoteConstants::NOTE_TO_BE_CREATED_UPDATED_AT_TIMESTAMP,
      user_id: UserConstants::USER_TO_BE_CREATED_ID

    created_note = NoteUseCases.get_note NoteConstants::NOTE_TO_BE_CREATED_ID

    assert_equal NoteConstants::NOTE_TO_BE_CREATED_ID, created_note.id
    assert_equal NoteConstants::NOTE_TO_BE_CREATED_TITLE, created_note.title
    assert_equal NoteConstants::NOTE_TO_BE_CREATED_BODY, created_note.body
    assert_equal NoteConstants::NOTE_TO_BE_CREATED_CREATED_AT_TIMESTAMP, created_note.created_at
    assert_equal NoteConstants::NOTE_TO_BE_CREATED_UPDATED_AT_TIMESTAMP, created_note.updated_at
  end

  test "Test Fetching All Notes From Database" do
    User.create id: UserConstants::USER_TO_BE_CREATED_ID, username: UserConstants::USER_TO_BE_CREATED_USERNAME

    Note.create id: NoteConstants::FIRST_NOTE_TO_BE_CREATED_ID,
      title: NoteConstants::NOTE_TO_BE_CREATED_TITLE,
      body: NoteConstants::NOTE_TO_BE_CREATED_BODY,
      created_at: NoteConstants::NOTE_TO_BE_CREATED_CREATED_AT_TIMESTAMP,
      updated_at: NoteConstants::NOTE_TO_BE_CREATED_UPDATED_AT_TIMESTAMP,
      user_id: UserConstants::USER_TO_BE_CREATED_ID

    Note.create id: NoteConstants::SECOND_NOTE_TO_BE_CREATED_ID,
      title: NoteConstants::NOTE_TO_BE_CREATED_TITLE,
      body: NoteConstants::NOTE_TO_BE_CREATED_BODY,
      created_at: NoteConstants::NOTE_TO_BE_CREATED_CREATED_AT_TIMESTAMP,
      updated_at: NoteConstants::NOTE_TO_BE_CREATED_UPDATED_AT_TIMESTAMP,
      user_id: UserConstants::USER_TO_BE_CREATED_ID

    created_notes = NoteUseCases.get_all_notes UserConstants::USER_TO_BE_CREATED_ID

    assert_equal NoteConstants::NUMBER_OF_NOTES_CREATED_BY_UNIQUE_USER, created_notes.length
  end

  test "Test Creating Note On Database" do
    User.create id: UserConstants::USER_TO_BE_CREATED_ID, username: UserConstants::USER_TO_BE_CREATED_USERNAME

    NoteUseCases.create_note NoteConstants::NOTE_TO_BE_CREATED_ID, NoteConstants::NOTE_TO_BE_CREATED_TITLE,
      NoteConstants::NOTE_TO_BE_CREATED_BODY,
      NoteConstants::NOTE_TO_BE_CREATED_CREATED_AT_TIMESTAMP,
      NoteConstants::NOTE_TO_BE_CREATED_UPDATED_AT_TIMESTAMP,
      UserConstants::USER_TO_BE_CREATED_ID

    created_note = Note.find NoteConstants::NOTE_TO_BE_CREATED_ID

    assert_equal NoteConstants::NOTE_TO_BE_CREATED_ID, created_note.id
    assert_equal NoteConstants::NOTE_TO_BE_CREATED_TITLE, created_note.title
    assert_equal NoteConstants::NOTE_TO_BE_CREATED_BODY, created_note.body
    assert_equal NoteConstants::NOTE_TO_BE_CREATED_CREATED_AT_TIMESTAMP, created_note.created_at
    assert_equal NoteConstants::NOTE_TO_BE_CREATED_UPDATED_AT_TIMESTAMP, created_note.updated_at
  end

  test "Test Updating Note On Database" do
    User.create id: UserConstants::USER_TO_BE_CREATED_ID, username: UserConstants::USER_TO_BE_CREATED_USERNAME

    Note.create id: NoteConstants::NOTE_TO_BE_CREATED_ID,
      title: NoteConstants::NOTE_TO_BE_CREATED_TITLE,
      body: NoteConstants::NOTE_TO_BE_CREATED_BODY,
      created_at: NoteConstants::NOTE_TO_BE_CREATED_CREATED_AT_TIMESTAMP,
      updated_at: NoteConstants::NOTE_TO_BE_CREATED_UPDATED_AT_TIMESTAMP,
      user_id: UserConstants::USER_TO_BE_CREATED_ID

    NoteUseCases.update_note NoteConstants::UPDATED_NOTE_TITLE,
      NoteConstants::UPDATED_NOTE_BODY,
      NoteConstants::NOTE_TO_BE_UPDATED_UPDATED_AT_TIMESTAMP,
      NoteConstants::NOTE_TO_BE_CREATED_ID

    updated_note = Note.find NoteConstants::NOTE_TO_BE_CREATED_ID

    assert_equal NoteConstants::UPDATED_NOTE_TITLE, updated_note.title
    assert_equal NoteConstants::UPDATED_NOTE_BODY, updated_note.body
    assert_equal NoteConstants::NOTE_TO_BE_UPDATED_UPDATED_AT_TIMESTAMP, updated_note.updated_at
  end

  test "Test Deleting Note On Database" do
    User.create id: UserConstants::USER_TO_BE_CREATED_ID, username: UserConstants::USER_TO_BE_CREATED_USERNAME

    Note.create id: NoteConstants::NOTE_TO_BE_CREATED_ID,
      title: NoteConstants::NOTE_TO_BE_CREATED_TITLE,
      body: NoteConstants::NOTE_TO_BE_CREATED_BODY,
      created_at: NoteConstants::NOTE_TO_BE_CREATED_CREATED_AT_TIMESTAMP,
      updated_at: NoteConstants::NOTE_TO_BE_CREATED_UPDATED_AT_TIMESTAMP,
      user_id: UserConstants::USER_TO_BE_CREATED_ID

    NoteUseCases.delete_note NoteConstants::NOTE_TO_BE_CREATED_ID

    assert_raises ActiveRecord::RecordNotFound do
      NoteUseCases.get_note NoteConstants::NOTE_TO_BE_CREATED_ID
    end
  end
end
